Add README.md file in .github directory

rwp0 opened the pull request:
Provide formatting with [Markdown](https://daringfireball.net/projects/markdown/) (as is suitable for GitHub repositories)

Hide the file in GitHub-specific [.github](https://github.com/Perl/perl5/tree/blead/.github) directory

Format code, and link to webpages where appropriate

-----

Improve [_main_](https://github.com/Perl/perl5/blob/blead/README) (first-seen) source code documentation readability,
HTML webpage formatting being very easy on readers' eyes

See also: [#17984](https://github.com/Perl/perl5/pull/17984) and #17227

rwp0 committed:
Add README.md file in .github directory

Provide formatting with Markdown

rwp0 committed:
Merge pull request #1 from rwp0/rwp0-readme.md

rwp0 committed:
format localperl directory

format localperl directory as code

rwp0 committed:
format list

format list ordered by letters

rwp0 committed:
correct links

provide full URL instead of realtive
as some links were pointing incorrectly

rwp0 committed:
Merge branch 'Perl:blead' into blead

nwc10 commented on .github/README.md:
```diff
@@ -0,0 +1,130 @@
+processing, and has one of the world's most impressive collections of
+third-party modules.
+
+For an introduction to the language's features, see [pod/perlintro.pod](https://perldoc.perl.org/perlhack).
+
```
As this is (now) the "README for github" file, I think that all these documentation links should point to github instead of perldoc.perl.org

Hence

    For an introduction to the language's features, see [pod/perlintro.pod](https://github.com/Perl/perl5/blob/blead/pod/perlhack.pod).

github renders the pod nicely, and in checking that, I realised that it actually makes **more** sense, given what `perlhack.pod` says:

```
=item * Ensure you're following the latest advice

In case the advice in this guide has been updated recently, read the
latest version directly from the perl source:

  % perldoc pod/perlhack.pod
```

perldoc.perl.org is only showing the version of the file from the most recent stable release, so might well be out of date. The blead blob is, well, the bleading edge.

rwp0 replied:
I agree, based on `perlhack` I'm going to modify URLs to link to more recent versions.

nwc10 commented on .github/README.md:
```diff
@@ -44,7 +42,7 @@ This will configure and compile perl for your platform, run the regression
 
 If you run into any trouble whatsoever or you need to install a customized
-version of Perl, you should read the detailed instructions in the "[INSTALL](INSTALL)"
+version of Perl, you should read the detailed instructions in the "[INSTALL](https://github.com/rwp0/perl5/blob/blead/INSTALL)"
```
You're linking to https://github.com/**rwp0**/perl5/blob/blead - whoops!

I think it would be better to take this commit and correct it with `git commit --amend` instead of pushing another fixup commit after it, as keeping two commits would just clutter the history.

(ie then update the branch with `git push --force-with-lease`)

(I wonder - can you use relative links? ie would ../INSTALL work here? Is the problem that this file is now down in `.github`? Relative links would be cleaner if they work correctly)

rwp0 replied:
Oops, right. Okay, I'll do so. Yes, sadly, the relative links I put before were pointing from `.github` directory.

rwp0 replied:
I'm not sure if I could use `git commit --amend` correctly, but changes are committed already.

nwc10 commented on .github/README.md:
````diff
@@ -41,7 +41,7 @@ install this version of Perl locally, run the following commands:
 ```
 
 This will configure and compile perl for your platform, run the regression
-tests, and install perl in a subdirectory "localperl" of your home directory.
````
I'm not sure if I'm nit-picking here - but I wondered whether we still need the "" quotes when the text is formatted in `code font`. The implication of the font is the same as the implication of the quotes - that this word is a literal token of some form.

I also wondered whether generated HTML page would read more cleanly if you remove the "" around filenames that are now links - ie

does this:

> you should read the detailed instructions in the [INSTALL](https://github.com/rwp0/perl5/blob/blead/INSTALL) file that came with this distribution

look better than this?

> you should read the detailed instructions in the "[INSTALL](https://github.com/rwp0/perl5/blob/blead/INSTALL)" file that came with this distribution

nwc10 replied:
Oops, I didn't manage to grab the next line. Your change was

```
 This will configure and compile perl for your platform, run the regression
-tests, and install perl in a subdirectory "localperl" of your home directory.
+tests, and install perl in a subdirectory "`localperl`" of your home directory.
```

Also, oops, github "helpfully" let me review an old commit, but I think that this is still the same change in the current commit.

rwp0 replied:
I also think so. Quotes are redundant in Markdown.

rwp0 committed:
fix documentation links to point to GitHub versions of Pod

remove double quotes from linked file and section names
amend some links to point to correct absolute path
format linked perldoc tool and documentation files as code

rwp0 commented:
Also I wonder if it is possible to somehow force GitHub (or Linguist) to render the extension-less `README.<OS>` (eg. [README.openbsd](https://github.com/Perl/perl5/blob/blead/README.openbsd)) files as POD (as they all are written in POD).

nwc10 commented:
> Also (on an unrelated note) I wonder if it is possible to somehow force GitHub (or Linguist) to render the extension-less `README.<OS>` (eg. [README.openbsd](https://github.com/Perl/perl5/blob/blead/README.openbsd)) files as POD (as they all contain and are written in POD) so that they can too be linked on GitHub, displayed formatted, and read comfortably.

By some amount of chance I ended up needing to look at Linguist this morning, and `.gitattributes`

I *think* that it can be done with a line in `.gitattributes` that sets `language=pod` for the files, using suitable file glob.

If you want to try this, could you do it in a separate branch and PR?

(and I might not get time to look this or your other changes today)

rwp0 commented:
> > Also (on an unrelated note) I wonder if it is possible to somehow force GitHub (or Linguist) to render the extension-less `README.<OS>` (eg. [README.openbsd](https://github.com/Perl/perl5/blob/blead/README.openbsd)) files as POD (as they all contain and are written in POD) so that they can too be linked on GitHub, displayed formatted, and read comfortably.
> 
> By some amount of chance I ended up needing to look at Linguist this morning, and `.gitattributes`
> 
> I _think_ that it can be done with a line in `.gitattributes` that sets `language=pod` for the files, using suitable file glob.
> 
> If you want to try this, could you do it in a separate branch and PR?
> 
> (and I might not get time to look this or your other changes today)

great, I gladly will work on it in a separate PR

nwc10 commented:
> I _think_ that it can be done with a line in `.gitattributes` that sets `language=pod` for the files, using suitable file glob.

I misremembered the details. It would need to be `linguist-language=pod` (if it works at all)

Docs are at https://github.com/github/linguist/blob/master/docs/overrides.md

Sorry - hope my mistake didn't waste your time.

haarg commented:
Setting linguist-language does not impact how GitHub renders documentation.

https://github.com/haarg/perl5/blob/ling-readme/.gitattributes
https://github.com/haarg/perl5/blob/ling-readme/README.macos

rwp0 commented:
> Setting linguist-language does not impact how GitHub renders documentation.
> 
> [haarg/perl5@`ling-readme`/.gitattributes](https://github.com/haarg/perl5/blob/ling-readme/.gitattributes) [haarg/perl5@`ling-readme`/README.macos](https://github.com/haarg/perl5/blob/ling-readme/README.macos)

https://github.com/rwp0/perl5/blob/readme.os-pod/.gitattributes

I also added `INSTALL` and `PACKAGING` files

btw the examples (in `override.md`) include to `-` (dash) before the option for specific files (i wonder why)

haarg commented:
linguist-language is used for syntax highlighting and classification for stats, not documentation rendering. It isn't a bug. It could be a feature request though.

rwp0 commented:
> linguist-language is used for syntax highlighting and classification for stats, not documentation rendering. It isn't a bug. It could be a feature request though.

ah yes, but the syntax for pod (as source) did not highlight with it :-)

rwp0 committed:
Merge branch 'Perl:blead' into blead

tom-binary commented:
Improving the README formatting is a good goal. However, maintaining two copies of README would be a shame - since Markdown is designed to be "almost" as readable as plain text, could we not just use this content in `README` directly? Or was that considered and rejected already?

(the specific changes such as having links to Github in the file seem reasonable enough, given that this is the primary version control for the project at the moment)

rwp0 commented:
> Improving the README formatting is a good goal. However, maintaining two copies of README would be a shame - since Markdown is designed to be "almost" as readable as plain text, could we not just use this content in `README` directly? Or was that considered and rejected already?
> 
> (the specific changes such as having links to Github in the file seem reasonable enough, given that this is the primary version control for the project at the moment)

I think having one traditional[^1] README for the source tarball, and one derived from it for GitHub is better idea.

As for the maintenance I consider it to be a static (documentation) file historically released with Perl which is not expected to change.

[^1]: as easily read on most standard text editors without much formatting

rwp0 committed:
Merge branch 'Perl:blead' into blead

rwp0 committed:
Merge branch 'Perl:blead' into blead

rwp0 committed:
Merge branch 'Perl:blead' into blead

rwp0 commented:
Why not merge this PR yet? 🤔

rwp0 committed:
Merge branch 'Perl:blead' into blead

rwp0 committed:
Merge branch 'Perl:blead' into blead

rwp0 committed:
Merge branch 'Perl:blead' into blead

rwp0 committed:
Merge branch 'Perl:blead' into blead

khwilliamson commented:
@rwp0 please rebase this so it can be safely merged

rwp0 committed:
Merge branch 'Perl:blead' into blead

rwp0 committed:
Capitalize Perl referring to the language.
